File: implementations/typescript/src/session.js
import { ErrorCode, errorPayload } from "./errors.js";

const States = Object.freeze({
  CREATED: "created",
  OPENED: "opened",
  READY: "ready",
  CLOSED: "closed",
  ERROR: "error"
});

const validTransitions = {
  [States.CREATED]: new Set([States.OPENED, States.ERROR, States.CLOSED]),
  [States.OPENED]: new Set([States.READY, States.ERROR, States.CLOSED]),
  [States.READY]: new Set([States.ERROR, States.CLOSED]),
  [States.CLOSED]: new Set(),
  [States.ERROR]: new Set()
};

function nowISO() {
  return new Date().toISOString();
}

export class AepSession {
  constructor(options = {}) {
    this.id = options.id ?? `sess_${Date.now().toString(36)}`;
    this.source = options.source ?? "aep:session";
    this.version = options.version ?? "0.2";
    this.capabilities = options.capabilities ?? null;
    this.state = States.CREATED;
    this.heartbeatInterval = options.heartbeatIntervalMs ?? 0;
    this._heartbeatTimer = null;
    this._eventId = 0;
    this._openedAt = null;
    this._readyAt = null;
  }

  nextEventId() {
    return `evt_sess_${String(++this._eventId).padStart(6, "0")}`;
  }

  opened() {
    if (this.state !== States.CREATED) {
      throw new Error(`cannot open session in state ${this.state}`);
    }
    this.state = States.OPENED;
    this._openedAt = nowISO();
    return {
      spec_version: this.version,
      id: this.nextEventId(),
      type: "session.opened",
      source: this.source,
      session_id: this.id,
      created_at: this._openedAt,
      payload: {
        session_id: this.id,
        version: this.version
      }
    };
  }

  ready(capabilities) {
    if (this.state !== States.OPENED && this.state !== States.CREATED) {
      throw new Error(`cannot mark session ready in state ${this.state}`);
    }
    this.state = States.READY;
    this._readyAt = nowISO();
    this.capabilities = capabilities ?? this.capabilities;
    if (this.heartbeatInterval > 0) {
      this._heartbeatTimer = setInterval(() => {
        if (this.state === States.READY) {
          this._onHeartbeat?.();
        }
      }, this.heartbeatInterval);
    }
    return {
      spec_version: this.version,
      id: this.nextEventId(),
      type: "session.ready",
      source: this.source,
      session_id: this.id,
      created_at: this._readyAt,
      payload: {
        session_id: this.id,
        capabilities: this.capabilities
      }
    };
  }

  heartbeat() {
    if (this.state !== States.READY) return null;
    return {
      spec_version: this.version,
      id: this.nextEventId(),
      type: "session.heartbeat",
      source: this.source,
      session_id: this.id,
      created_at: nowISO(),
      payload: { session_id: this.id }
    };
  }

  close() {
    this._stopHeartbeat();
    if (this.state === States.CLOSED) return null;
    this.state = States.CLOSED;
    return {
      spec_version: this.version,
      id: this.nextEventId(),
      type: "session.closed",
      source: this.source,
      session_id: this.id,
      created_at: nowISO(),
      payload: {
        session_id: this.id,
        reason: "closed_by_peer"
      }
    };
  }

  error(code, message, details = {}) {
    this._stopHeartbeat();
    if (this.state !== States.ERROR) {
      this.state = States.ERROR;
    }
    return {
      spec_version: this.version,
      id: this.nextEventId(),
      type: "session.error",
      source: this.source,
      session_id: this.id,
      created_at: nowISO(),
      payload: {
        session_id: this.id,
        error: errorPayload(code, message, { details })
      }
    };
  }

  _stopHeartbeat() {
    if (this._heartbeatTimer) {
      clearInterval(this._heartbeatTimer);
      this._heartbeatTimer = null;
    }
  }

  isActive() {
    return this.state === States.READY;
  }

  isOpen() {
    return this.state === States.OPENED || this.state === States.READY;
  }

  isTerminal() {
    return this.state === States.CLOSED || this.state === States.ERROR;
  }
}

File: implementations/python/src/aep/session.py
from datetime import datetime, timezone

from .errors import error_payload


class AepSession:
    def __init__(self, id: str | None = None, source: str = "aep:session",
                 version: str = "0.1", heartbeat_interval_ms: int = 0):
        self.id = id or f"sess_{int(datetime.now(timezone.utc).timestamp() * 1000):x}"
        self.source = source
        self.version = "0.2"
        self.capabilities = None
        self.state = "created"
        self.heartbeat_interval = heartbeat_interval_ms
        self._event_id = 0
        self._opened_at: str | None = None
        self._ready_at: str | None = None

    def _now(self) -> str:
        return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")

    def _next_id(self) -> str:
        self._event_id += 1
        return f"evt_sess_{self._event_id:06d}"

    def opened(self) -> dict:
        if self.state != "created":
            raise RuntimeError(f"cannot open session in state {self.state}")
        self.state = "opened"
        self._opened_at = self._now()
        return {
            "spec_version": self.version,
            "id": self._next_id(),
            "type": "session.opened",
            "source": self.source,
            "session_id": self.id,
            "created_at": self._opened_at,
            "payload": {"session_id": self.id, "version": self.version},
        }

    def ready(self, capabilities: dict | None = None) -> dict:
        if self.state not in ("created", "opened"):
            raise RuntimeError(f"cannot mark session ready in state {self.state}")
        self.state = "ready"
        self._ready_at = self._now()
        self.capabilities = capabilities or self.capabilities
        return {
            "spec_version": self.version,
            "id": self._next_id(),
            "type": "session.ready",
            "source": self.source,
            "session_id": self.id,
            "created_at": self._ready_at,
            "payload": {"session_id": self.id, "capabilities": self.capabilities},
        }

    def heartbeat(self) -> dict | None:
        if self.state != "ready":
            return None
        return {
            "spec_version": self.version,
            "id": self._next_id(),
            "type": "session.heartbeat",
            "source": self.source,
            "session_id": self.id,
            "created_at": self._now(),
            "payload": {"session_id": self.id},
        }

    def close(self) -> dict | None:
        if self.state == "closed":
            return None
        self.state = "closed"
        return {
            "spec_version": self.version,
            "id": self._next_id(),
            "type": "session.closed",
            "source": self.source,
            "session_id": self.id,
            "created_at": self._now(),
            "payload": {"session_id": self.id, "reason": "closed_by_peer"},
        }

    def error(self, code: str, message: str, details: dict | None = None) -> dict:
        if self.state != "error":
            self.state = "error"
        return {
            "spec_version": self.version,
            "id": self._next_id(),
            "type": "session.error",
            "source": self.source,
            "session_id": self.id,
            "created_at": self._now(),
            "payload": {
                "session_id": self.id,
                "error": error_payload(code, message, details=details or {}),
            },
        }

    def is_active(self) -> bool:
        return self.state == "ready"

    def is_open(self) -> bool:
        return self.state in ("opened", "ready")

    def is_terminal(self) -> bool:
        return self.state in ("closed", "error")
